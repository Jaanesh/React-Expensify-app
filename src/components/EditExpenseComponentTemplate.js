import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseFormComponent';
import {editExpense,removeExpense,startRemoveExpense} from '../actions/expense.js';

/*
const editExpenseComponentTemplate=(props)=>{    
    return(
    <div>
        <h2>Edit the Expense of {props.expenseToBeEdited.id} </h2>
        <br/>
        <ExpenseForm
              expenseToBeEdited={props.expenseToBeEdited}    
              onSubmit={(updatedExpense)=>{
                console.log(updatedExpense);
                props.dispatch(editExpense(props.expenseToBeEdited.id,updatedExpense));
                props.history.push('/');
            }}              
              >
        </ExpenseForm>  

        <br/><br/>
                <button onClick={()=>{
                                         console.log('remove id='+props.expenseToBeEdited.id);
                                         props.dispatch(removeExpense(props.expenseToBeEdited.id));
                                         props.history.push('/');
                                     }} 
                                    >Remove</button> 
    </div>
    )    
};

const mapStateToProps=(state,props)=>{
      return {
          expenseToBeEdited:state.expenses.find((eachExpense)=>{
               // console.log('eachExpense.id='+eachExpense.id);
               // console.log('props.match.params.id='+props.match.params.id);
                 return eachExpense.id===props.match.params.id ;
          })
          //we dont requie filter properties here as we need only existing expenses to edit
          //filters:state.filters
      };
  }
export default connect(mapStateToProps)(editExpenseComponentTemplate);
*/

export class EditExpenseComponentTemplate extends React.Component{

    onSubmit=(updatedExpense)=>{
        console.log(updatedExpense);
        this.props.updateExpense(this.props.expenseToBeEdited.id,updatedExpense);
        this.props.history.push('/');
    }

    onClick=()=>{
        console.log('remove id='+this.props.expenseToBeEdited.id);
        this.props.removeExistingExpense(this.props.expenseToBeEdited.id);
        this.props.history.push('/');
    }

    render(){
        return(
            <div>
                <h2>Edit the Expense of {this.props.expenseToBeEdited.id} </h2>
                <br/>
                <ExpenseForm
                      expenseToBeEdited={this.props.expenseToBeEdited}    
                      onSubmit={this.onSubmit}              
                      >
                </ExpenseForm>  
        
                <br/><br/>
                        <button onClick={this.onClick}>Remove</button> 
            </div>
            )    
    }
}

const mapStateToProps=(state,props)=>{
    return {
        expenseToBeEdited:state.expenses.find((eachExpense)=>{           
               return eachExpense.id===props.match.params.id ;
        })       
    };
}

const mapDispatchToProps=(dispatch,props)=>{
   // return onSubmit=(newExpense)=>dispatch(addExpense(newExpense));
     return{
        updateExpense:(id,updatedExpense)=>dispatch(editExpense(id,updatedExpense)),
        //removeExistingExpense:(id)=>dispatch(removeExpense(id))
        removeExistingExpense:(id)=>dispatch(startRemoveExpense(id))
        
     }
      
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpenseComponentTemplate);