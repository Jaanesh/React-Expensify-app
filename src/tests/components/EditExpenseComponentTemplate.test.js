import React from 'react';
import {shallow} from 'enzyme';
import moment from 'moment';
import {EditExpenseComponentTemplate} from '../../components/EditExpenseComponentTemplate.js';

let expenseObj1={id:'1', description:'jan rent',note:'',amount:3,createdAt:moment(0).
                                                                    subtract(5,'days').valueOf()};

let updateobj={id:'2', description:'Feb rent',note:'',amount:3,createdAt:moment(0).
                                                                    subtract(5,'days').valueOf()};
let wrapper,updateExpense,removeExistingExpense,history,expenseToBeEdited,onSubmit,onClick;

beforeEach(()=>{    
    updateExpense=jest.fn();
    removeExistingExpense=jest.fn();
    history={push:jest.fn()};
    expenseToBeEdited=expenseObj1;
    wrapper=shallow(<EditExpenseComponentTemplate expenseToBeEdited={expenseToBeEdited} 
                                                  updateExpense={updateExpense}
                                                  removeExistingExpense={removeExistingExpense}
                                                  history={history} />);
});

test('should render edit expense',()=>{
         expect(wrapper).toMatchSnapshot();
});

test('should handle edit expene',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(updateobj);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(updateExpense).toHaveBeenLastCalledWith('1',updateobj);
});

test('should handle remove expense',()=>{
    wrapper.find('button').prop('onClick')();
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExistingExpense).toHaveBeenLastCalledWith('1');
});