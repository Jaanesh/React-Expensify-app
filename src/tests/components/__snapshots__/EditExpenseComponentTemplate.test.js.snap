// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense 1`] = `
<div>
  <h2>
    Edit the Expense of 
    1
     
  </h2>
  <br />
  <ExpenseForm
    expenseToBeEdited={
      Object {
        "amount": 3,
        "createdAt": -432000000,
        "description": "jan rent",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <br />
  <br />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
